src/audio/multiband_drc/multiband_drc.c

```diff
@@ -0,0 +1,605 @@
+#include <user/eq.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_multiband_drc;
+
+/* 0d9f2256-8e4f-47b3-8448-239a334f1191 */
+DECLARE_SOF_RT_UUID("multiband_drc", multiband_drc_uuid, 0x0d9f2256, 0x8e4f, 0x47b3,
+		    0x84, 0x48, 0x23, 0x9a, 0x33, 0x4f, 0x11, 0x91);
+
+DECLARE_TR_CTX(multiband_drc_tr, SOF_UUID(multiband_drc_uuid), LOG_LEVEL_INFO);
+
+static inline void multiband_drc_crossover_reset_state(struct multiband_drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_ch(&state->crossover[i]);
+}
+
+static inline void multiband_drc_emphasis_reset_state(struct multiband_drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_lr4(&state->emphasis[i]);
+}
+
+static inline void multiband_drc_deemphasis_reset_state(struct multiband_drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_lr4(&state->deemphasis[i]);
+}
+
+static inline void multiband_drc_drc_reset_state(struct multiband_drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < SOF_MULTIBAND_DRC_MAX_BANDS; i++)
+		drc_reset_state(&state->drc[i]);
+}
+
+static inline void multiband_drc_reset_state(struct multiband_drc_state *state)
+{
+	/* Reset emphasis eq-iir state */
+	multiband_drc_emphasis_reset_state(state);
+
+	/* Reset crossover state */
+	multiband_drc_crossover_reset_state(state);
+
+	/* Reset drc kernel state */
+	multiband_drc_drc_reset_state(state);
+
+	/* Reset deemphasis eq-iir state */
+	multiband_drc_deemphasis_reset_state(state);
+}
+
+static int multiband_drc_eq_init_coef_ch(struct sof_eq_iir_biquad_df2t *coef,
+					 struct iir_state_df2t *eq)
+{
+	int ret;
+
+	eq->coef = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			   sizeof(struct sof_eq_iir_biquad_df2t) * 2);
+	if (!eq->coef)
+		return -ENOMEM;
+
+	/* Coefficients of the first biquad and second biquad */
+	ret = memcpy_s(eq->coef, sizeof(struct sof_eq_iir_biquad_df2t) * 2,
+		       coef, sizeof(struct sof_eq_iir_biquad_df2t) * 2);
+	assert(!ret);
+
+	/* EQ filters are two 2nd order filters, so only need 4 delay slots
+	 * delay[0..1] -> state for first biquad
+	 * delay[2..3] -> state for second biquad
+	 */
+	eq->delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			    sizeof(uint64_t) * CROSSOVER_NUM_DELAYS_LR4);
+	if (!eq->delay)
+		return -ENOMEM;
+
+	eq->biquads = 2;
+	eq->biquads_in_series = 2;
+
+	return 0;
+}
+
+static int multiband_drc_crossover_eq_init_coef(struct multiband_drc_comp_data *cd, int16_t nch)
+{
+	struct sof_eq_iir_biquad_df2t *crossover;
+	struct sof_eq_iir_biquad_df2t *emphasis;
+	struct sof_eq_iir_biquad_df2t *deemphasis;
+	struct sof_multiband_drc_config *config = cd->config;
+	struct multiband_drc_state *state = &cd->state;
+	int ch, err;
+
+	if (!config) {
+		comp_cl_err(&comp_multiband_drc, "crossover_eq_init_coef(), no config is set");
```

Comment:
Crossover_eq_init_coef() is no more the function name, these trace prints should be updated. The actual function name is pretty long too, maybe a shorter would do?

Reply:
The convention of the function is <component_name>_<function>. multiband_drc is long already, maybe use mbdrc_<function> instead?

Reply:
Followed cujomalainey's comment I collapsed the *_init_coef functions into one, called multiband_drc_init_coef. It should look better now.